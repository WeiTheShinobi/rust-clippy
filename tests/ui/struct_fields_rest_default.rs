//@aux-build:proc_macros.rs
#![warn(clippy::struct_fields_rest_default)]
extern crate proc_macros;

#[derive(Default)]
struct Foo {
    a: i32,
    b: i32,
    c: i32,
}

impl Foo {
    fn get_foo() -> Self {
        Foo { a: 0, b: 0, c: 0 }
    }
}

fn main() {
    #[rustfmt::skip]
    let _ = Foo {
        a: 10,
        ..Default::default()
        //~^ struct_fields_rest_default
    };

    #[rustfmt::skip]
    let _ = Foo {
        a: 10,
        ..Foo::default()
        //~^ struct_fields_rest_default
    };

    // should not lint
    #[rustfmt::skip]
    let _ = Foo {
        a: 10,
        ..Foo::get_foo()
    };

    // should not lint in external macro
    proc_macros::external! {
        #[derive(Default)]
        struct ExternalDefault {
            a: i32,
            b: i32,
        }

        let _ = ExternalDefault {
            a: 10,
            ..Default::default()
        };
    }

    // should not lint
    let _ = Foo {
        a: Default::default(),
        b: Default::default(),
        c: Default::default(),
    };
}
