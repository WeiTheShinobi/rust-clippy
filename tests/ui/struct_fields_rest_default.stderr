error: should not use `..Default::default()` to omit rest of struct field initialization
  --> tests/ui/struct_fields_rest_default.rs:22:11
   |
LL |         ..Default::default()
   |           ^^^^^^^^^^^^^^^^^^
   |
help: each field's initial value should be explicitly specified
  --> tests/ui/struct_fields_rest_default.rs:20:13
   |
LL |       let _ = Foo {
   |  _____________^
LL | |         a: 10,
LL | |         ..Default::default()
LL | |
LL | |     };
   | |_____^
   = note: `-D clippy::struct-fields-rest-default` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::struct_fields_rest_default)]`

error: should not use `..Foo::default()` to omit rest of struct field initialization
  --> tests/ui/struct_fields_rest_default.rs:29:11
   |
LL |         ..Foo::default()
   |           ^^^^^^^^^^^^^^
   |
help: each field's initial value should be explicitly specified
  --> tests/ui/struct_fields_rest_default.rs:27:13
   |
LL |       let _ = Foo {
   |  _____________^
LL | |         a: 10,
LL | |         ..Foo::default()
LL | |
LL | |     };
   | |_____^

error: should not use `..Foo::get_foo()` to omit rest of struct field initialization
  --> tests/ui/struct_fields_rest_default.rs:36:11
   |
LL |         ..Foo::get_foo()
   |           ^^^^^^^^^^^^^^
   |
help: each field's initial value should be explicitly specified
  --> tests/ui/struct_fields_rest_default.rs:34:13
   |
LL |       let _ = Foo {
   |  _____________^
LL | |         a: 10,
LL | |         ..Foo::get_foo()
LL | |
LL | |     };
   | |_____^

error: aborting due to 3 previous errors

